warp10/src/main/java/io/warp10/standalone/datalog/DatalogWorker.java

```diff
@@ -0,0 +1,123 @@
+import io.warp10.continuum.gts.GTSEncoder;
+import io.warp10.continuum.gts.GTSWrapperHelper;
+import io.warp10.continuum.sensision.SensisionConstants;
+import io.warp10.continuum.store.StoreClient;
+import io.warp10.continuum.store.thrift.data.DatalogRecord;
+import io.warp10.continuum.store.thrift.data.DatalogRecordType;
+import io.warp10.continuum.store.thrift.data.Metadata;
+import io.warp10.sensision.Sensision;
+import io.warp10.standalone.StandaloneDirectoryClient;
+import io.warp10.standalone.datalog.DatalogWorkers.DatalogJob;
+
+public class DatalogWorker extends Thread {
+
+  private final LinkedBlockingQueue<DatalogJob> queue;
+
+  public DatalogWorker(LinkedBlockingQueue<DatalogJob> queue) {
+    this.queue = queue;
+
+    this.setName("[Datalog Worker]");
+    this.setDaemon(true);
+    this.start();
+  }
+
+  @Override
+  public void run() {
+    StoreClient storeClient = DatalogWorkers.getStoreClient();
+    StandaloneDirectoryClient directoryClient = (StandaloneDirectoryClient) DatalogWorkers.getDirectoryClient();
+
+    DatalogManager manager = null;
+    if (storeClient instanceof DatalogStoreClient) {
+      manager = ((DatalogStoreClient) storeClient).getDatalogManager();
+    }
+    boolean hasManager = null != manager;
+
+    Map<String,String> labels = new LinkedHashMap<String,String>();
+
+    while(true) {
+
+      DatalogJob job = null;
+      DatalogRecord record = null;
+
+      Throwable err = null;
+
+      try {
+        job = queue.poll();
+
+        if (null == job) {
+          LockSupport.parkNanos(10000000L);
+          continue;
+        }
+
+        record = job.record;
+
+        System.out.println(job.record);
+
+        if (hasManager) {
+          manager.process(record);
+          job.consumer.success(job.ref);
+        } else {
+          Metadata metadata = record.getMetadata();
+          switch (job.record.getType()) {
+            case UPDATE:
+              GTSEncoder enc = null;
+              GTSDecoder decoder = new GTSDecoder(record.getBaseTimestamp(), record.bufferForEncoder());
+              decoder.next();
+              enc = decoder.getEncoder();
+              storeClient.store(enc);
+              job.consumer.success(job.ref);
+              break;
+            case REGISTER:
+              directoryClient.register(metadata);
+              job.consumer.success(job.ref);
+              break;
+            case UNREGISTER:
+              directoryClient.unregister(metadata);
+              job.consumer.success(job.ref);
+              break;
+            case DELETE:
+              storeClient.delete(null, metadata, record.getStart(), record.getStop());
+              job.consumer.success(job.ref);
+              break;
+          }
+        }
+      } catch (Throwable t) {
+        err = t;
+        if (null != job) {
+          job.consumer.failure(job.ref);
+        }
+      } finally {
+        labels.put(SensisionConstants.SENSISION_LABEL_TYPE, job.record.getType().name());
```

Comment:
This could lead to NullPointerException.
There are some cases where job is null or job.record.getType() is null.

Reply:
When can job be null? There is an explicit conditional that continues the loop if job is null. When can job ne non null and type be null?

Reply:
Solved, the job is null because the call to continue will first go through the finally block. For the type it was a concurrency issue which has been fixed in commit 677fd1e.